README
Sailthru::Client

DOCUMENTATION

http://getstarted.sailthru.com/new-for-developers-overview/api-client-library/perl

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

Digest::MD5
JSON::XS
LWP::Protocol::https
LWP::UserAgent
Params::Validate
Readonly
URI
Test::MockModule
Test::Exception

COPYRIGHT AND LICENSE

Copyright (C) 2012 by Finn Smith <finn@timeghost.net>

Copyright (C) 2012 by Steve Sanbeg <stevesanbeg@buzzfeed.com>

Copyright (C) 2011 by Steve Miketa <steve@sailthru.com>

Adapted from the original Triggermail module created by Sam Gerstenzang.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl module Sailthru::Client.

2.1.0 Wed May 4th 2016
    - Added support to get Rate Limiting information

2.0.0 Thu Sep 13 14:57:07 2012
    - Major refactoring to match other Sailthru libs.
    - Deprecation of old method calls.
    - Increased and improved test coverage.

1.2.0 Wed Aug 8 14:47:23 2012
    - Adding support for UTF8 chars going to the md5_hex function

1.1.0 Fri Dec 16 14:47:23 2011
    - Adding support for UTF8 chars going to the md5_hex function

1.0.0 Fri Nov 18 10:28:00 2011
    - Repackaging as Sailthru::Client for official package support by Sailthru

1.0.4 Thu Nov 17 11:38:00 2011
    - Updating _flatten_hash to build appropriate html parameter arrays

1.0.3 Mon Aug  2 11:54:01 2011
    - Use https to connect to Sailthru API instead of http.

1.0.2 Mon Jul 18 14:52:12 2011
    - Bugfix: don't die when receiving non-200 status code response from Sailthru API.

1.0.1 Wed Apr 27 13:15:38 2011
    - Fixed tests to take API key and secret as environment variables.

1.0.0 Fri Mar  4 15:57:26 2011
    - Patch added by David Yee from 20x200. Now supports schedule_time in send method.

0.2.0 Fri Jul  9 15:57:26 2010
    - Made checking more lenient, added copyTemplate.

0.1.0 Thu Jul  1 11:57:26 2010
    - original version; created by h2xs 1.23 with options
        -XA -n Sailthru


lib/Triggermail.pm
package Triggermail;

our $VERSION = '1.003';

use strict;
use warnings;

use constant API_URI => 'https://api.sailthru.com';

use Digest::MD5 qw(md5_hex);
use JSON::XS;
use LWP;
use HTTP::Request;
use URI::Escape;
use Params::Validate qw(:all);

sub new {
	my $class = shift;
	my $self  = {
		api_key => shift,
		secret  => shift,
		timeout => shift,
	};

	return bless $self, $class;
}

sub getEmail {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR } );
	my ( $self, $email ) = @_;
	my %data = ( email => $email );
	return $self->_apiCall( 'email', \%data, 'GET' );
}

sub setEmail {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR }, 0, 0, 0 );
	my ( $self, $email, $vars_ref, $lists_ref, $templates_ref ) = @_;
	my %data;
	$data{'email'} = $email;
	$self->_flatten_hash( 'vars',  $vars_ref,  \%data ) if $vars_ref;
	$self->_flatten_hash( 'lists', $lists_ref, \%data ) if $lists_ref;
	$self->_flatten_hash( 'templates', $templates_ref, \%data )
	  if $templates_ref;
	return $self->_apiCall( 'email', \%data, 'POST' );
}

sub send {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR }, { type => SCALAR }, 0, 0, 0 );
	my %data;
	my ( $self, $template, $email, $vars_hash, $options_hash, $schedule_time ) = @_;
	$data{'template'}      = $template;
	$data{'email'}         = $email;
	$data{'schedule_time'} = $schedule_time if $schedule_time;
	$self->_flatten_hash( 'vars',    $vars_hash,    \%data ) if $vars_hash;
	$self->_flatten_hash( 'options', $options_hash, \%data ) if $options_hash;
	return $self->_apiCall( 'send', \%data, 'POST' );
}

sub getSend {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR } );
	my ( $self, $send_id ) = @_;
	my %data = ( send_id => $send_id );
	return $self->_apiCall( 'send', \%data, 'GET' );
}

sub scheduleBlast {
	validate_pos(
		@_,
		{ type => HASHREF },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		0
	);
	my ( $self, $name, $list, $schedule_time, $from_name, $from_email, $subject, $content_html, $content_text,
		$options ) = @_;
	my %data = (
		name          => $name,
		list          => $list,
		schedule_time => $schedule_time,
		from_name     => $from_name,
		from_email    => $from_email,
		subject       => $subject,
		content_html  => $content_html,
		content_text  => $content_text
	);
	if ($options) {
		my %merged_hash = ( %data, %{$options} );    #merge in the options hash
		%data = %merged_hash;
	}
	return $self->_apiCall( 'blast', \%data, 'POST' );
}

sub getBlast {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR } );
	my ( $self, $blast_id ) = @_;
	my %data = ( blast_id => $blast_id );
	return $self->_apiCall( 'blast', \%data, 'GET' );
}

sub copyTemplate {
	validate_pos(
		@_,
		{ type => HASHREF },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		{ type => SCALAR },
		0
	);
	my ( $self, $template, $data_feed, $setup, $subject_line, $schedule_time, $list, $options ) = @_;
	my %data = (
		copy_template => $template,
		data_feed_url => $data_feed,
		setup         => $setup,
		name          => $subject_line,
		schedule_time => $schedule_time,
		list          => $list,
	);

	# $self->_flatten_hash( 'options', $options, \%data ) if $options;
	if ($options) {
		my %merged_hash = ( %data, %{$options} );    #merge in the options hash
		%data = %merged_hash;
	}
	return $self->_apiCall( 'blast', \%data, 'POST' );
}

sub getTemplate {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR } );
	my ( $self, $template ) = @_;
	my %data = ( template => $template );
	return $self->_apiCall( 'template', \%data, 'GET' );
}

sub importContacts {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR }, 0 );
	my ( $self, $email, $password, $include_names ) = @_;
	$include_names = 0 if ( !$include_names );
	my %data = (
		email         => $email,
		password      => $password,
		include_names => $include_names
	);
	return $self->_apiCall( 'contacts', \%data, 'POST' );
}

sub _apiCall {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR }, { type => HASHREF }, { type => SCALAR } );

	my ( $self, $action, $data, $method ) = @_;

	$data->{'api_key'} = $self->{api_key};
	$data->{'format'}  = 'json';
	$data->{'sig'}     = $self->_getSignatureHash($data);
	my $result = $self->_httpRequest( API_URI . "/" . $action, $data, $method );
	my $json    = JSON::XS->new->ascii->pretty->allow_nonref;
	my $decoded = $json->decode( $result->content );
	return $decoded ? $decoded : $result;
}

sub _httpRequest {
	validate_pos( @_, { type => HASHREF }, { type => SCALAR }, { type => HASHREF }, { type => SCALAR } );

	my ( $self, $url, $data, $method ) = @_;
	my $browser = LWP::UserAgent->new;
	$browser->timeout( $self->{timeout} ) if $self->{timeout};
	my $response;
	if ( $method eq 'POST' ) {
		$response = $browser->post( $url, $data );
	}
	else {    #GET
		use URI;
		$url = URI->new($url);
		$url->query_form( %{$data} );
		$response = $browser->get($url);
	}

# don't die on based on status code: other sailthru clients don't
#    die "$url error: ", $response->status_line
#      unless $response->is_success;

	if ($response) {
		return $response;
	}
}

sub _getSignatureHash {
	validate_pos( @_, { type => HASHREF }, { type => HASHREF } );

	my ( $self, $params ) = @_;
	my @values;
	$self->_extractValues( $params, \@values );
	@values = sort @values;
	my $string = $self->{secret} . join( '', @values );
	return md5_hex($string);
}

#sub _flatten_hash {
#	validate_pos( @_, { type => HASHREF }, { type => SCALAR }, { type => HASHREF }, { type => HASHREF } );
#	my ( $self, $name, $nested_hash, $mother_hash ) = @_;
#	while ( ( my $key, my $value ) = each %{$nested_hash} ) {
#		if (   ref( $nested_hash->{$key} ) eq 'HASH'
#			|| ref( $nested_hash->{$key} ) eq 'REF' ) {
#			$self->_flatten_hash( $key, $nested_hash->{$key}, $mother_hash );
#		}
#		else {
#			$mother_hash->{ $name . "[" . $key . "]" } = $value;
#		}
#	}
#}

sub _flatten_hash {
    validate_pos(
        @_,
        { type => HASHREF },
        { type => SCALAR },
        { type => HASHREF },
        { type => HASHREF },
    ); 
    my ( $self, $name, $nested_hash, $mother_hash ) = @_;
    my @parents;
   
    $self->_flatten_hash_routine( $nested_hash, $mother_hash, \@parents);
}


sub _flatten_hash_routine {

    validate_pos(
        @_,
        { type => HASHREF },
        { type => HASHREF },
        { type => HASHREF },
        { type => ARRAYREF }
    ); 

    my ($self, $unflattened, $flattened, $parents) = @_;
    while (my ($key, $value) = each(%$unflattened)) {
        push (@$parents, $key);
        my $type_name = ref($value);

        if ($type_name eq "HASH") {
            $self->_flatten_hash_routine($value, $flattened, $parents);   

        } elsif ($type_name eq "ARRAY") {
            my $parent_idx = 0;
            foreach (@$value) {
                my $array_val = $_;
                push (@$parents, $parent_idx++);
                my $array_idx = $self->_build_url_idx($parents);
                $flattened->{$array_idx} = $array_val;
                pop(@$parents);
            }  
        } else  {
            my $array_idx = $self->_build_url_idx($parents);
            $flattened->{$array_idx} = $value;
        }  
        pop (@$parents);
    }  
}

sub _build_url_idx {
    validate_pos(
        @_,
        { type => HASHREF },
        { type => ARRAYREF }
    );
    my ($self, $dimension) = @_;
    return  'vars[' . join('][', @$dimension ) . ']';
}


sub _extractValues {
	validate_pos( @_, { type => HASHREF }, { type => HASHREF }, { type => ARRAYREF } );
	my ( $self, $hash, $array ) = @_;
	while ( ( my $key, my $value ) = each %{$hash} ) {
		if ( ref($value) eq 'HASH' || ref($value) eq 'REF' ) {
			$self->_extractValues( $value, $array );
		}
		else {
			push @{$array}, $value;
		}
	}
}

1;
__END__

=head1 NAME

Triggermail - Perl module for accessing SailThru's platform

=head1 SYNOPSIS

 use Triggermail;
 my $tm = Triggermail->new('api_key','secret'); # You can optionally include a timeout in seconds as a third parameter.
 %vars = ( name => "Joe Example", from_email => "approved_email@your_domain.com", your_variable => "some_value");
 %options = ( reply_to => "your reply_to header");
 $tm->send("template_name",'example@example.com',\%vars,\%options);

=head1 DESCRIPTION

Triggermail is a Perl module for accesing the Sailthru platform.

All methods return a hash with return values. Dump the hash or explore the SailThru API documentation page for what might be returned.

L<http://docs.sailthru.com/api>

Some options might change. Always consult the SailThru API documentation for the best information.

=head2 METHODS

=over 4

=item C<getEmail( $email, )>

=item C<setEmail( $email, \%vars, \%lists, \%templates )>

Takes email as string. vars, lists, templates as hash references.
The vars hash you choose your own key/values for later substitution.
The lists hash should be of format list_name => 1 for subscribed, 0 for unsubscribed.
The templates hash is a list of templates user has opted out, use the key as the template name to signal opt-out.
As always, see the Sailthru documentation for more information.

=item C<send( $template, $email, \%vars, \%options, $schedule_time )>

Send an email to a single address.
Takes template, email and schedule_time as strings. vars, options as hash references.

Options:

=over

=item C<replyto>

override Reply-To header

=item C<test>

send as test email (subject line will be marked, will not count towards stats)

=back

=item C<getSend( $send_id )>

Check if send worked, using send_id returned in the hash from send()

=item C<scheduleBlast( $name, $list, $schedule_time, $from_name, $from_email, $subject, $content_html, $content_text, \%options )>

Schedule an email blast. See the API documentation for more details on what should be passed.

L<http://docs.sailthru.com/api/blast>

=item C<getBlast( $blast_id )>

Check if blast worked, using blast_id returned in the hash from scheduleBlast()
Takes blast_id.

=item C<copyTemplate( $template_name, $data_feed, $setup, $subject_line, $schedule_time, $list )>

Allows you to use an existing template to send out a blast.

=item C<getTemplate( $template_name )>

Retrieves information about the template

=item C<importContacts( $email, $password )>

Import contacts from major providers.
Takes email, password as strings. By default does not include names. Pass 1 as third argument to include names.

=back


=head1 SEE ALSO

See the SailThru API documentation for more details on their API.

L<http://docs.sailthru.com/api>

=head1 AUTHOR

Sam Gerstenzang

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sam Gerstenzang

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/Sailthru/Client.pm
package Sailthru::Client;

use strict;
use warnings;

use Carp;
use JSON::XS;
use LWP::UserAgent;
use Digest::MD5 qw( md5_hex );
use Params::Validate qw( :all );
use Readonly;
use URI;

our $VERSION = '2.1.0';
Readonly my $API_URI => 'https://api.sailthru.com/';

#
# public api
#

# args:
#
# * api_key - scalar
# * secret - scalar
# * timeout - scalar (optional)
sub new {
    my ( $class, $api_key, $secret, $timeout ) = @_;
    my $self = {
        api_key => $api_key,
        secret  => $secret,
        ua      => LWP::UserAgent->new,
        last_rate_limit_info_ref => {}
    };
    $self->{ua}->timeout($timeout) if $timeout;
    $self->{ua}->default_header( 'User-Agent' => "Sailthru API Perl Client $VERSION" );
    return bless $self, $class;
}

# args:
#
# * template_name - scalar
# * email - scalar
# * vars - hashref (optional)
# * options - hashref (optional)
# * schedule_time - scalar (optional)
sub send {
    my $self   = shift;
    my @params = validate_pos(
        @_,
        { type => SCALAR },
        { type => SCALAR },
        { type => HASHREF, default => {} },
        { type => HASHREF, default => {} },
        { type => SCALAR, default => undef }
    );
    my ( $template_name, $email, $vars, $options, $schedule_time ) = @params;
    my $data = {};
    $data->{template}      = $template_name;
    $data->{email}         = $email;
    $data->{vars}          = $vars if keys %{$vars};
    $data->{options}       = $options if keys %{$options};
    $data->{schedule_time} = $schedule_time if $schedule_time;
    return $self->api_post( 'send', $data );
}

# args:
# * send_id - scalar
sub get_send {
    my $self      = shift;
    my @params    = validate_pos( @_, { type => SCALAR } );
    my ($send_id) = @params;
    return $self->api_get( 'send', { send_id => $send_id } );
}

# args:
# * email - scalar
sub get_email {
    my $self    = shift;
    my @params  = validate_pos( @_, { type => SCALAR } );
    my ($email) = @params;
    return $self->api_get( 'email', { email => $email } );
}

# args:
# * email - scalar
# * vars - hashref (optional)
# * lists - hashref (optional)
# * templates - hashref (optional)
sub set_email {
    my $self   = shift;
    my @params = validate_pos(
        @_,
        { type => SCALAR },
        { type => HASHREF, default => {} },
        { type => HASHREF, default => {} },
        { type => HASHREF, default => {} }
    );
    my ( $email, $vars, $lists, $templates ) = @params;
    my $data = {};
    $data->{email}     = $email;
    $data->{vars}      = $vars if keys %{$vars};
    $data->{lists}     = $lists if keys %{$lists};
    $data->{templates} = $templates if keys %{$templates};
    return $self->api_post( 'email', $data );
}

# args:
# * name - scalar
# * list - scalar
# * schedule_time - scalar
# * from_name - scalar
# * from_email - scalar
# * subject - scalar
# * content_html - scalar
# * content_text - scalar
# * options - hashref (optional)
sub schedule_blast {
    my $self   = shift;
    my @params = validate_pos(
        @_,
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => HASHREF, default => {} }
    );
    my ( $name, $list, $schedule_time, $from_name, $from_email, $subject, $content_html, $content_text, $options ) =
      @params;
    # initialize our data hash by copying the contents of the options hash
    my $data = { %{$options} };
    $data->{name}          = $name;
    $data->{list}          = $list;
    $data->{schedule_time} = $schedule_time;
    $data->{from_name}     = $from_name;
    $data->{from_email}    = $from_email;
    $data->{subject}       = $subject;
    $data->{content_html}  = $content_html;
    $data->{content_text}  = $content_text;
    return $self->api_post( 'blast', $data );
}

# args:
# * template_name - scalar
# * list - scalar
# * schedule_time - scalar
# * options - hashref (optional)
sub schedule_blast_from_template {
    my $self   = shift;
    my @params = validate_pos(
        @_,
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => HASHREF, default => {} },
    );
    my ( $template_name, $list, $schedule_time, $options ) = @params;
    # initialize our data hash by copying the contents of the options hash
    my $data = { %{$options} };
    $data->{copy_template} = $template_name;
    $data->{list}          = $list;
    $data->{schedule_time} = $schedule_time;
    return $self->api_post( 'blast', $data );

}

# args:
# * blast_id - scalar
sub get_blast {
    my $self       = shift;
    my @params     = validate_pos( @_, { type => SCALAR } );
    my ($blast_id) = @params;
    return $self->api_get( 'blast', { blast_id => $blast_id } );
}

# args:
# * template_name - scalar
sub get_template {
    my $self            = shift;
    my @params          = validate_pos( @_, { type => SCALAR } );
    my ($template_name) = @params;
    return $self->api_get( 'template', { template => $template_name } );
}

# args:
# * action - scalar
# * data - hashref
sub api_get {
    my $self = shift;
    my @params = validate_pos( @_, { type => SCALAR }, { type => HASHREF } );
    my ( $action, $data ) = @params;
    return $self->_api_request( $action, $data, 'GET' );
}

# args:
# * action - scalar
# * data - hashref
# * TODO: optional binary_key arg
sub api_post {
    my $self = shift;
    my @params = validate_pos( @_, { type => SCALAR }, { type => HASHREF } );
    my ( $action, $data ) = @params;
    return $self->_api_request( $action, $data, 'POST' );
}

# args:
# * action - scalar
# * data - hashref
sub api_delete {
    my $self = shift;
    my @params = validate_pos( @_, { type => SCALAR }, { type => HASHREF } );
    my ( $action, $data ) = @params;
    return $self->_api_request( $action, $data, 'DELETE' );
}

# args:
#
# * action - scalar
# * method - scalar
sub get_last_rate_limit_info {
    my $self   = shift;
    my @params = validate_pos(
        @_,
        { type => SCALAR },
        { type => SCALAR }
    );
    my ( $action, $method ) = @params;
    $method = uc($method);

    if (exists $self->{last_rate_limit_info_ref}->{$action} and $self->{last_rate_limit_info_ref}->{$action}->{$method}) {
        return $self->{last_rate_limit_info_ref}->{$action}->{$method};
    }

    return;
}

#
# private helper methods
#

# args:
# * action - scalar
# * data - hashref
# * request_type - scalar
sub _api_request {
    my $self = shift;
    my @params = validate_pos( @_, { type => SCALAR }, { type => HASHREF }, { type => SCALAR } );
    my ( $action, $data, $request_type ) = @params;
    my $payload    = $self->_prepare_json_payload($data);
    my $action_uri = $API_URI . $action;
    my $response   = $self->_http_request( $action_uri, $payload, $request_type );

    # update rate limit information
    if ( defined $response->header('x-rate-limit-limit') and
         defined $response->header('x-rate-limit-remaining') and
         defined $response->header('x-rate-limit-reset') ) {
         $self->{last_rate_limit_info_ref}->{$action}->{$request_type} = {
             limit => $response->header('x-rate-limit-limit'),
             remaining => $response->header('x-rate-limit-remaining'),
             reset => $response->header('x-rate-limit-reset')
         };
    }

    return decode_json( $response->content );
}

# args:
# * uri - scalar
# * data - hashref
# * method - scalar
sub _http_request {
    my $self = shift;
    my @params = validate_pos( @_, { type => SCALAR }, { type => HASHREF }, { type => SCALAR } );
    my ( $uri, $data, $method ) = @params;
    $uri = URI->new($uri);
    my $response;
    if ( $method eq 'GET' ) {
        $uri->query_form($data);
        $response = $self->{ua}->get($uri);
    }
    elsif ( $method eq 'POST' ) {
        $response = $self->{ua}->post( $uri, $data );
    }
    elsif ( $method eq 'DELETE' ) {
        $uri->query_form($data);
        $response = $self->{ua}->delete($uri);
    }
    else {
        croak "Invalid method: $method";
    }
    return $response;
}

# args:
# * data - hashref
sub _prepare_json_payload {
    my $self    = shift;
    my @params  = validate_pos( @_, { type => HASHREF } );
    my ($data)  = @params;
    my $payload = {};
    $payload->{api_key} = $self->{api_key};
    $payload->{format}  = 'json';
    # this gives us nice clean utf8 encoded json text
    $payload->{json} = encode_json($data);
    $payload->{sig} = $self->_get_signature_hash( $payload, $self->{secret} );
    return $payload;
}

# Every request must also generate a signature hash called sig according to the
# following rules:
#
# * take the string values of every parameter, including api_key
# * sort the values alphabetically, case-sensitively (i.e. ordered by Unicode code point)
# * concatenate the sorted values, and prepend this string with your shared secret
# * generate an MD5 hash of this string and use this as sig
# * now generate your URL-encoded query string from your parameters plus sig

# args:
# * params - hashref
# * secret - scalar
# NOTE This internal method assumes a single level hash with values for only 'api_key', 'format', and 'json'
# NOTE Since we pack everything into the 'json' value this is safe and we do not need to recurse down a nested hash.
sub _get_signature_hash {
    my $self = shift;
    my @params = validate_pos( @_, { type => HASHREF }, { type => SCALAR } );
    my ( $api_param_hash, $secret ) = @params;
    my @api_param_values = values %{$api_param_hash};
    my $sig_string = join '', $secret, sort @api_param_values;
    # assumes utf8 encoded text, works fine because we use encode_json internally
    return md5_hex($sig_string);
}

### XXX
### DEPRECATED METHODS
### XXX

# args:
# * email - scalar
sub getEmail {
    my $self = shift;
    warnings::warnif( 'deprecated', 'getEmail is deprecated, use get_email instead' );
    return $self->get_email(@_);
}

# args:
# * email - scalar
# * vars - hashref (optional)
# * lists - hashref (optional)
# * templates - hashref (optional)
sub setEmail {
    my $self = shift;
    warnings::warnif( 'deprecated', 'setEmail is deprecated, use set_email instead' );
    return $self->set_email(@_);
}

# args:
# * send_id - scalar
sub getSend {
    my $self = shift;
    warnings::warnif( 'deprecated', 'getSend is deprecated, use get_send instead' );
    return $self->get_send(@_);
}

# args:
# * name - scalar
# * list - scalar
# * schedule_time - scalar
# * from_name - scalar
# * from_email - scalar
# * subject - scalar
# * content_html - scalar
# * content_text - scalar
# * options - hashref (optional)
sub scheduleBlast {
    my $self = shift;
    warnings::warnif( 'deprecated', 'scheduleBlast is deprecated, use schedule_blast instead' );
    return $self->schedule_blast(@_);
}

# args:
# * blast_id - scalar
sub getBlast {
    my $self = shift;
    warnings::warnif( 'deprecated', 'getBlast is deprecated, use get_blast instead' );
    return $self->get_blast(@_);
}

sub copyTemplate {
    my $self   = shift;
    my @params = validate_pos(
        @_,
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => SCALAR },
        { type => HASHREF, default => {} }
    );
    my ( $template, $data_feed, $setup, $subject_line, $schedule_time, $list, $options ) = @params;
    warnings::warnif( 'deprecated', 'copyTemplate is deprecated, use schedule_blast_from_template instead' );
    # initialize our data hash by copying the contents of the options hash
    my $data = { %{$options} };
    $data->{copy_template} = $template;
    $data->{data_feed_url} = $data_feed;
    $data->{setup}         = $setup;
    $data->{name}          = $subject_line;
    $data->{schedule_time} = $schedule_time;
    $data->{list}          = $list;
    return $self->api_post( 'blast', $data );
}

# args:
# * template_name - scalar
sub getTemplate {
    my $self = shift;
    warnings::warnif( 'deprecated', 'getTemplate is deprecated, use get_template instead' );
    return $self->get_template(@_);
}

# args:
# * email - scalar
# * password - scalar
# * include_names - scalar (optional)
sub importContacts {
    my $self = shift;
    my @p = validate_pos( @_, { type => SCALAR }, { type => SCALAR }, { type => SCALAR, default => 0 } );
    my ( $email, $password, $include_names ) = @p;
    warnings::warnif( 'deprecated',
        'importContacts is deprecated. The contacts API has been discontinued as of August 1st, 2011.' );
    my $data = {
        email         => $email,
        password      => $password,
        include_names => $include_names,
    };
    return $self->api_post( 'contacts', $data );
}

1;

__END__

=head1 NAME

Sailthru::Client - Perl module for accessing Sailthru's API

=head1 SYNOPSIS

    use Sailthru::Client;

    # instantiate a new Sailthru::Client with an api_key and secret
    $sc = Sailthru::Client->new('api_key', 'secret');

    # send an email to a single email address
    %vars = (
       name => "Joe Example",
       from_email => "approved_email@your_domain.com",
       your_variable => "some_value"
    );
    %options = ( reply_to => "your reply_to header");
    $sc->send('template_name', 'example@example.com', \%vars, \%options);

=head1 DESCRIPTION

Sailthru::Client is a Perl module for accessing the Sailthru API.

Methods return a reference to a hash containing the response values. Dump the
hash or read the Sailthru API documentation for which values are returned
by which API calls.

L<http://docs.sailthru.com/api>

Some options might change. Consult the Sailthru API documentation for
the latest information.

=head1 METHODS

=head2 Sailthru::Client->new( $api_key, $secret, [$timeout] )

Returns a new Sailthru::Client object.

=over

=item $api_key

Sailthru API key.

=item $secret

Sailthru API secret.

=item $timeout

Optional network timeout in seconds.

=back

=head2 $sc->send( $template_name, $email, [\%vars, \%options, $schedule_time] )

Remotely send an email template to a single email address.

API docs: L<http://docs.sailthru.com/api/send>

=over

=item $template_name

The name of the template to send.

=item $email

The email address to send to.

=item \%vars

An optional hashref of the replacement vars to use in the send. Each var may be referenced as {varname} within the template itself.

=item \%options

An optional hashref to include a replyto header, test keys, etc. See the API documentation for details.

=item $schedule_time

Do not send the email immediately, but at some point in the future. Any date recognized by PHP's strtotime function is valid, but be sure to specify timezone or use a UTC time to avoid confusion. You may also use relative time.

=back

=head2 $sc->get_send( $send_id )

Get the status of a send.

API docs: L<http://docs.sailthru.com/api/send>

=over

=item $send_id

The unique identifier of the send returned in the response from C<$sc-E<gt>send()>.

=back

=head2 $sc->get_email( $email )

Get information about a user.

API docs: L<http://docs.sailthru.com/api/email>

=over

=item $email

The email address to look up.

=back

=head2 $sc->set_email( $email, [\%vars, \%lists, \%templates] )

Update information about a user, including adding and removing the user from lists.

API docs: L<http://docs.sailthru.com/api/email>

=over

=item $email

The email address to modify.

=item \%vars

An optional hashref of replacement variables you want to set or a JSON string

=item \%lists

An optional hashref. Each key is the name of a list and each value is
1 to subscribe the user to that list or 0 to remove the user from the list.

=item \%templates

An optional hashref. Each key is the name of a template, and each value is 1 to
opt the user back in to template delivery or 0 to opt the user out of template
delivery.

=back

=head2 $sc->schedule_blast( $name, $list, $schedule_time, $from_name, $from_email, $subject, $content_html, $content_text, [\%options] )

Schedule a mass mail blast.

API docs: L<http://docs.sailthru.com/api/blast>

=over

=item $name

The name to give to this new blast.

=item $list

The mailing list name to send to.

=item $schedule_time

When the blast should send. Dates in the past will be scheduled for immediate
delivery. Any English textual datetime format known to PHP's strtotime function
is acceptable, such as 2012-03-18 23:57:22 UTC, now (immediate delivery), +3
hours (3 hours from now), or March 18, 9:30 EST. Be sure to specify a timezone
if you use an exact time.

=item $from_name

The name to use in "From" in the email.

=item $from_email

The email address to use in "From". Choose from any of your verified emails.

=item $subject

The subject line of the email.

=item $content_html

The HTML format version of the email.

=item $content_text

The text format version of the email.

=item \%options

An optional hashref containing the optional parameters for a blast. See the API documentation for details.

=back

=head2 $sc->schedule_blast_from_template( $template_name, $list, $schedule_time, [\%options] )

Schedule a mass mail blast from a template.

API docs: L<http://docs.sailthru.com/api/blast>

=over

=item $template_name

The template to copy from.

=item $list

The mailing list name to send to.

=item $schedule_time

When the blast should send. Dates in the past will be scheduled for immediate
delivery. Any English textual datetime format known to PHP's strtotime function
is acceptable, such as 2012-03-18 23:57:22 UTC, now (immediate delivery), +3
hours (3 hours from now), or March 18, 9:30 EST. Be sure to specify a timezone
if you use an exact time.

=item \%options

An optional hashref containing the optional parameters for a blast. See the API documentation for details.

=back

=head2 $sc->get_blast( $blast_id )

Get data on a single blast.

API docs: L<http://docs.sailthru.com/api/blast>

=over

=item $blast_id

The blast id returned in the response from C<$sc-E<gt>scheduleBlast()>.

=back

=head2 $sc->get_template( $template_name )

Get information about a template

API docs: L<http://docs.sailthru.com/api/template>

=over

=item $template_name

The name of the template.

=back

=head2 $sc->api_get( $action, \%data )

This is a generic HTTP GET call to the API.

=over

=item $action

The name of the API action to call.

=item \%data

A hashref of arguments to pass to the API.

=back

For example, you could get information about an email with

 $sc->api_get('GET', 'email', {email=>'somebody@example.com'});

=head2 $sc->api_post( $action, \%data )

This is a generic HTTP POST call to the API.

=over

=item $action

The name of the API action to call.

=item \%data

A hashref of arguments to pass to the API.

=back

=head2 $sc->api_delete( $action, \%data )

This is a generic HTTP DELETE call to the API.

=over

=item $action

The name of the API action to call.

=item \%data

A hashref of arguments to pass to the API.

=back

=head2 $sc->get_last_rate_limit_info( $action, $method )

Get the last rate limit information .

=over

=item $action

The name of the API action to call.

=item $method

Http request type. One of GET, POST or DELETE.

=item return

Hash reference with three fields (limit, remaining and reset) or undef if not exists.

=back

=head1 SEE ALSO

See the Sailthru API documentation for more details on their API.

L<http://docs.sailthru.com/api>

=head1 AUTHOR

Finn Smith

Steve Sanbeg

Steve Miketa

Sam Gerstenzang

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2012 by Finn Smith <finn@timeghost.net>

Copyright (C) 2012 by Steve Sanbeg <stevesanbeg@buzzfeed.com>

Copyright (C) 2011 by Steve Miketa <steve@sailthru.com>

Adapted from the original Sailthru::Client & Triggermail modules created by Sam
Gerstenzang and Steve Miketa.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself, either Perl version 5.10.0 or, at your option,
any later version of Perl 5 you may have available.

=cut
